### DISCUSSION TITLE
PostgreSQL migration error after adding and removing fields

### DISCUSSION THREAD

**Discussion by jschuur** [prisma/prisma#10759] [Q&A] [answered]
I'm upgrading Next Auth to 4.x and it requires some [schema changes](https://next-auth.js.org/getting-started/upgrade-v4#schema-changes) to the tables it uses (which are managed via @next-auth/prisma-adapter).

I went ahead and updated my `schema.prisma` by removing/adding/renaming the fields as indicated by the upgrade guide, but get the following error now when running `prisma migrate dev` (prisma 3.6.0 and PostgreSQL 14.1):

```
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Datasource "db": PostgreSQL database "streamersdev", schema "public" at "localhost:5432"

⚠️  Warnings for the current datasource:

  • A unique constraint covering the columns `[provider,providerAccountId]` on the table `Account` will be added. If there are existing duplicate values, this will fail.

✔ Are you sure you want create and apply this migration? … yes
✔ Enter a name for the new migration: … nextauth4_migration
Applying migration `20211218183954_nextauth4_migration`
Error: P3018

A migration failed to apply. New migrations cannot be applied before the error is recovered from. Read more about how to resolve migration issues in a production database: https://pris.ly/d/migrate-resolve

Migration name: 20211218183954_nextauth4_migration

Database error code: 42601

Database error:
ERROR: syntax error at or near ","

Position:
 47 ALTER TABLE "Session" DROP COLUMN "accessToken",
 48 DROP COLUMN "createdAt",
 49 DROP COLUMN "updatedAt";
 50
 51 -- AlterTable
 52 ALTER TABLE "User" RENAME CONSTRAINT "User_pkey1" TO "User_pkey",

DbError { severity: "ERROR", parsed_severity: Some(Error), code: SqlState("42601"), message: "syntax error at or near \",\"", detail: None, hint: None, position: Some(Original(2681)), where_: None, schema: None, table: None, column: None, datatype: None, constraint: None, file: Some("scan.l"), line: Some(1180), routine: Some("scanner_yyerror") }
```

The schema.prisma changes come down to this:

```diff
model Account {
   id                 String    @id @default(cuid())
   userId             String
-  providerType       String
-  providerId         String
+  type               String
+  provider           String
   providerAccountId  String
   refreshToken       String?
   accessToken        String?
-  accessTokenExpires DateTime?
-  createdAt          DateTime  @default(now())
-  updatedAt          DateTime  @updatedAt
-  user               User      @relation(fields: [userId], references: [id])
-
-  @@unique([providerId, providerAccountId])
+  expires_in         DateTime?
+  expires_at         Int?
+  token_type         String?
+  scope              String?
+  id_token           String?
+  session_state      String?
+  oauth_token_secret String?
+  oauth_token        String?
+
+  user               User      @relation(fields: [userId], references: [id], onDelete: Cascade)
+
+  @@unique([provider, providerAccountId])
 }
 
 model Session {
   id           String   @id @default(cuid())
+  sessionToken String   @unique
   userId       String
   expires      DateTime
-  sessionToken String   @unique
-  accessToken  String   @unique
-  createdAt    DateTime @default(now())
-  updatedAt    DateTime @updatedAt
-  user         User     @relation(fields: [userId], references: [id])
+  user         User     @relation(fields: [userId], references: [id], onDelete: Cascade)
 }
 
 model User {
@@ -147,20 +150,15 @@ model User {
   email         String?   @unique
   emailVerified DateTime?
   image         String?
-  createdAt     DateTime  @default(now())
-  updatedAt     DateTime  @updatedAt
   accounts      Account[]
   sessions      Session[]
   isAdmin       Boolean   @default(false)
 }

-model VerificationRequest {
-  id         String   @id @default(cuid())
+model VerificationTokens {
   identifier String
   token      String   @unique
   expires    DateTime
-  createdAt  DateTime @default(now())
-  updatedAt  DateTime @updatedAt
 
   @@unique([identifier, token])
 }
```

It is perhaps worth noting that I periodically copy my production database over to my local machine via pg_dump and pg_restore. The production database is hosted on AWS Lightsail via the following npm script:

```
    "db:pull": "npm run db:dump:prod && npm run db:import:prod"
    "db:dump:prod": "dotenv -e .env.production -- pg_dump -Fc -f tmp/dbdump_prod_latest.dump",
    "db:import:prod": "pg_restore --no-acl --no-owner --clean -d streamersdev tmp/dbdump_prod_latest.dump",
```

Is what I am doing too complex for a Prisma migration to handle and I need to do this manually? Or is this down to how I am copying the database over from production?

**Comment by jschuur** [comment, marked as answer]
I ran into this issue even with a fresh (reset, empty), local database, when trying to add a different new model, and still got a weird migration error, so I've filed a bug report at #10769 instead.

In fact, this seems to be an issue to do with a named constrain bug posted at @ #8940 recently.

**Reply by nikolasburk** [reply]
Hey Joost, thanks for raising this issue! It seems like you've already resolved your issue now, correct? 

By the way, I want to throw out a quick compliment for the clear write-up of your problem, proactively filing a bug report and keeping the information you shared on GitHub up-to-date! 💚 

![image](https://user-images.githubusercontent.com/4058327/146754635-fd191a11-cfcc-4bb9-bbac-463a619565bb.png)

**Reply by jschuur** [reply]
Thanks for checking @nikolasburk, and the kind words about the issue too. I can confirm that my issue in this discussion thread was indeed resolved once I first ran `prisma migrate dev` without any `schema.prisma` changes like I mentioned in the issue/screenshot above.

Everything good on production too.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
